### DISCUSSION TITLE
Title bar and caption controls (minimize, maximize & close buttons)

### DISCUSSION THREAD

**Discussion by Craig-K2** [dotnet/maui#8228] [Q&A]
We have a requirement for an app that when started, shouldn't be closed through the user interface. Yes its an anit-pattern but due to the critical nature of the app it must remain running and ready to delegate messages.

Using MAUI the app displays 2 different windows concurrently - targeting Windows 10, 11 & macCatalyst

How can I;
1. Have a window that has no title bar visible or caption controls
2. Disable/hide the maximize & close buttons on the other window but leave the minimize button enabled and retain this functionailty
3. Prevent both windows from being resized
4. Set a window to always on top

Have tried several angle of attack including handler mappings and ConfigureLifecycleEvents without success as MAUI overrides or completly ignores. The AppWindow presenter when it is a OverLappedPresenter seems to have all the properties I'm looking for but none seem to do anything concrete when set. 

Any solutions or guidenance much appreciated

**Comment by progmars** [comment]
Joining in with the same question but with a bit different context and some fruitless experiments I've tried.

I want to make my app only minimizable but not resizable and not maximizable. Ideally both on Windows and Mac, but I wanted to get it working on Windows first.

Based on examples, I tried the following approaches.

Platforms/Windows/App.xaml.cs:

    public App()
    {
        this.InitializeComponent();
        Microsoft.Maui.Handlers.WindowHandler.Mapper.AppendToMapping(nameof(IWindow), (handler, view) =>
        {
            var nativeWindow = handler.PlatformView;
            nativeWindow.Activate();
            IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(nativeWindow);
            WindowId windowId = Win32Interop.GetWindowIdFromWindow(hWnd);
            AppWindow appWindow = AppWindow.GetFromWindowId(windowId);

            var presenter = appWindow.Presenter as OverlappedPresenter;
            presenter.IsResizable = false;
            presenter.IsMaximizable = false;
        });
    }

IsResizable works, now borders cannot be resized. However, the Maximize button and control menu Maximize command is still present.

Tried the same code in an overridden method:

    protected override void OnLaunched(LaunchActivatedEventArgs args)
    {
        base.OnLaunched(args);

        var currentWindow = Application.Windows[0].Handler.PlatformView;
        IntPtr _windowHandle = WindowNative.GetWindowHandle(currentWindow);
        var windowId = Win32Interop.GetWindowIdFromWindow(_windowHandle);

        AppWindow appWindow = AppWindow.GetFromWindowId(windowId);
        var presenter = appWindow.Presenter as OverlappedPresenter;
        presenter.IsResizable = false;
        presenter.IsMaximizable = false;
        presenter.IsMinimizable = false;
    }

The same result, only `IsResizable = false;` has an effect.

Then I tried to move the same code to MauiProgram.cs:

    public static class MauiProgram
    {
        public static MauiApp CreateMauiApp()
        {
            var builder = MauiApp.CreateBuilder();
            builder
                .UseMauiApp<App>()
                .ConfigureFonts(fonts =>
                {
                    fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
                    fonts.AddFont("OpenSans-Semibold.ttf", "OpenSansSemibold");
                });

    #if WINDOWS
            builder.ConfigureLifecycleEvents(events =>
            {
                events.AddWindows(wndLifeCycleBuilder =>
                {
                    wndLifeCycleBuilder.OnWindowCreated(window =>
                    {
                        IntPtr nativeWindowHandle = WinRT.Interop.WindowNative.GetWindowHandle(window);
                        WindowId win32WindowsId = Win32Interop.GetWindowIdFromWindow(nativeWindowHandle);
                        AppWindow winuiAppWindow = AppWindow.GetFromWindowId(win32WindowsId);

                        var p = winuiAppWindow.Presenter as OverlappedPresenter;
                        p.IsResizable = false;
                        p.IsMaximizable = false;
                    });
                });
            });
    #endif     

The same result, only `IsResizable = false;` has an effect.

I even tried messing with Interop and SetWindowLong:

    flags &= ~(int)(WS.SizeFrame | WS.MaximizeBox);
    SetWindowLongPtr(hWnd, GWL_STYLE, (IntPtr)flags);

which worked fine for WinForms apps (although not needed there because you can configure it in Properties). But MAUI window seems to ignoring all my attempts.

How do I make maximize go away? Why MAUI is not honoring the flags that worked fine in WPF? None of the titlebar settings seem to work.

**Reply by Mathyn** [reply]
Did you ever manage to find a solution for this? I'm seeing the same behavior here and am struggling to find a solution.

**Reply by e012345678** [reply]
Same problem here, any update?

**Comment by codechavez** [comment]
I'm running into the same problem, I have tried the same code you have tried.  This is obviously an issue. Please Fix!

**Comment by Craig-K2** [comment]
As our project was green field,  we’ve had to change some of the window behaviour due to the restrictions placed on the MAUI window container running on Windows and macCatalyst. I can imagine this will be a deal breaker for those wanting to migrate over to MAUI without incurring loss of this key windowing functionality.

Did however manage to force one of my windows to be on top by implementing 
presenter.IsAlwaysOnTop = true; in a AppendToMapping method.

As well as preventing the closure of a window by hooking into the OnClosed

#if WINDOWS
                    events.AddWindows(windows => windows
                          .OnClosed((window, args) => args.Handled = true)
...
#endif


It would be nice to hear from the MAUI team. Find out what there plans are or better still - get a working solution from them :)

**Reply by yuzd** [reply]
this issue still exists in .net7
```csharp
p.IsMaximizable = false;// not work
```

**Reply by datadumped** [reply]
The code below works for me to hide the minimize and maximize button. Disable the close button, still visible but does not close the app when clicked. Using .Net 7 in Maui for windows app.

#if WINDOWS
using Microsoft.UI;
using Microsoft.UI.Windowing;
using Windows.Graphics;
using Microsoft.Maui.LifecycleEvents;
#endif

#if WINDOWS
            builder.ConfigureLifecycleEvents(events =>
            {
                events.AddWindows(wndLifeCycleBuilder =>
                {
                    wndLifeCycleBuilder.OnWindowCreated(window =>
                    {

                        IntPtr nativeWindowHandle = WinRT.Interop.WindowNative.GetWindowHandle(window);
                        WindowId win32WindowsId = Win32Interop.GetWindowIdFromWindow(nativeWindowHandle);
                        AppWindow winuiAppWindow = AppWindow.GetFromWindowId(win32WindowsId);  
                    
                        //hide the minimize and maximize button in title bar
                        window.ExtendsContentIntoTitleBar = false; 
                    

                        if(winuiAppWindow.Presenter is OverlappedPresenter p)
                        { 
                           p.Maximize();
                           p.IsAlwaysOnTop=true;
                           p.IsResizable=false;
                           p.IsMaximizable = false;
                           p.IsMinimizable=false;
                           p.IsAlwaysOnTop=true;

                           //disable the close button action when clicked - close button still visible, but does not close the app
                           events.AddWindows(windows => windows
                               .OnClosed((window, args) => args.Handled = true));
                        }                     
                        else
                        {
                            const int width = 1920;
                            const int height = 1080;
                            winuiAppWindow.MoveAndResize(new RectInt32(1920 / 2 - width / 2, 1080 / 2 - height / 2, width, height));                      
                        }                        
                    });
                });
            });
            #endif

**Reply by yuzd** [reply]
nice ~~~

**Reply by progmars** [reply]
@datadumped  Won't `window.ExtendsContentIntoTitleBar = false;` switch the title to the old Windows mode where you cannot have custom controls in the titlebar?

**Reply by green-munkey** [reply]
First of all, thanks for the code snippet. We had a requirement of removing the title bar and buttons completely. 
Below is a snippet that worked for me to achieve this as the original snippet was still showing the close button.

Project:
.NET 6
.NET MAUI Blazor

Below is an adjusted code snippet that did work for me. 

```
                             wndLifeCycleBuilder.OnWindowCreated(window =>
				{

	                                IntPtr nativeWindowHandle = WinRT.Interop.WindowNative.GetWindowHandle(window);
					WindowId win32WindowsId = Win32Interop.GetWindowIdFromWindow(nativeWindowHandle);
					AppWindow winuiAppWindow = AppWindow.GetFromWindowId(win32WindowsId);

					window.ExtendsContentIntoTitleBar = false; // must be false or else you see some of the buttons
					winuiAppWindow.SetPresenter(AppWindowPresenterKind.FullScreen);

				});

```

**Reply by datadumped** [reply]
@green-munkey Thanks for sharing! This is exactly what I've been looking for!

**Reply by ataparia** [reply]
Thanks @green-munkey . You saved my day.

**Comment by sllanesmanuel** [comment]
i have a problem, when i try to copy and paste the code that you put here it dont works i dont really know why, i put the code in another pc and it works fine :) 
![Imagen de WhatsApp 2023-10-07 a las 19 14 29_99fc5398](https://github.com/dotnet/maui/assets/85197668/4dd8132e-280c-4d45-ba7e-de536faa0785)
